lib/compress.js

```diff
@@ -171,8 +172,7 @@ merge(Compressor.prototype, {
             if (compressor.has_directive("use asm")) return self;
             var opt = optimizer(self, compressor);
             opt._optimized = true;
-            if (opt === self) return opt;
-            return opt.transform(compressor);
+            return opt;
```

Comment:
Problem with calling `opt.transform(compressor)` here is that the stack would change from:
```
...
...
self
```
to:
```
...
...
self
opt
```
which will cause optimisations that depend knowing where they are in the AST to fail.

The solution is to wait for `opt` [to be assigned on the relevant field](https://github.com/mishoo/UglifyJS2/blob/f4c2ea37bf9231b6f76804e74ee157be916280de/lib/transform.js#L57-L87) in the parent `AST_Node`, then `descend()` from the parent again. As this node is already marked as `_optimized`, only its children be altered by `optimize()`.